src/tiled/addremoveterrain.cpp

```diff
@@ -58,6 +58,22 @@ void AddRemoveTerrain::addTerrain()
     mTerrain = nullptr;
 }
 
+void AddRemoveTerrain::moveTerrainUp()
+{
+    Q_ASSERT(mTerrain);
```

Comment:
You're initializing the terrain to `nullptr` from the constructor of `MoveTerrainUp` and `MoveTerrainDown`, so this assertion failed for me. Should probably be `Q_ASSERT(!mTerrain);`? And then you can also get rid of the `mTerrain = nullptr;` below.

However, I wonder why these classes are derived from `AddRemoveTerrain`. What is the benefit of doing that? I'd suggest to make a separate `MoveTerrain` base class for these two instead.

Reply:
I've fixed this by creating a MoveTerrain class that only has the necessary members for moving terrains.